Sends Reportback from Phoenix API -> Rogue

chloealee opened the pull request:
#### What's this PR do?
- When an app hits Phoenix's `/api/v1/campaigns/[nid]/reportback` endpoint, instead of saving directly to Phoenix, it will send along to Rogue first and then Rogue will send back to Phoenix to be saved.
- If crop values aren't set, don't send to Rogue (both on first send and on cron job).

#### How should this be reviewed?
- Hit the `/api/v1/campaigns/[nid]/reportback` endpoint.
![screen shot 2016-11-17 at 3 51 11 pm](https://cloud.githubusercontent.com/assets/9019452/20407050/b646eb72-acdd-11e6-837f-9aa246408c0b.png)
- Check the Rogue `reportback_items` table for the reportback item.
- Check the Phoenix `dosomething_reportback_file` table for the reportback item. 
- Check that Phoenix's `dosoemthing_rogue_reportbacks` table has the reportback item with all correct data.

#### Relevant tickets
Fixes #7183 

#### Checklist
- [ ] Documentation added for new features/changed endpoints.
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love
- [ ] Post a message in #phoenix if this includes something that causes a rebuild!

chloealee committed:
adds ternary for cropping values and comments out code

chloealee committed:
adds todos and comments back in things we need

chloealee committed:
bookmark

chloealee committed:
works but cant connect to rogue - makes it into the failed logs table tho

chloealee committed:
this works if you have cropping values in the data

chloealee committed:
continue

chloealee committed:
only sends cropping if it is present

chloealee committed:
updates cron job - only send cropping values if they are set

chloealee committed:
updates cron job - only send cropping values if they are set

chloealee committed:
updates cron job - only send cropping values if they are set

chloealee committed:
adds space back in

chloealee committed:
WIP

chloealee committed:
changes  to

chloealee committed:
everything works

chloealee committed:
adds logic to save it in dosomething_rogue_reportbacks table

chloealee commented:
Dropping this here before I forget! Before I merge, should I notify anyone about this? I think Aaron is out so not sure who else to mention it to?

angaither commented on lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:
```diff
@@ -352,6 +352,25 @@ function _campaign_resource_reportback($nid, $values) {
-  return dosomething_reportback_save($values, $user);
+  $transaction_id = $_SERVER['HTTP_X_REQUEST_ID'];
+
+  if (! $transaction_id) {
```
can you add some comments here to explain why we are only sending the reportback if there isn't a transaction id

chloealee replied:
yes good call!

angaither commented on lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module:
```diff
@@ -71,14 +70,17 @@ function dosomething_rogue_send_reportback_to_rogue($values, $user = NULL) {
     'source' => isset($values['source']) ? $values['source'] : NULL,
   ];
 
+  if ($values['crop_x']) {
```
since we are changing this code in multiple places it might make sense to pull this out into a helper function that prepares the item before sending to rouge, but you can do a follow up PR after merge if you want

chloealee replied:
ah yes, great call! Since the two are structured a bit differently (ie `$data['crop_x']` vs `$values['crop_x']` and are set to an array vs object `$values['crop_x']` vs `$task->crop_x`) would it be cleanest to write a helper function with conditionals checking if it's in the `$data` format, set to array, else set to the `$task` object?

angaither replied:
or cast the array as an object when you pass it in so you always are dealing with the same format

chloealee replied:
ah i see, thank you!

chloealee replied:
@angaither this isn't actually working - I will tag you when it is!

chloealee replied:
@angaither I'm having trouble wrapping my head around this for some reason, can I do this in a separate PR like you suggested above in order to not hold up this PR from being merged?

chloealee replied:
actually I think I got something to work!

chloealee replied:
@angaither I'm playing around with this and think it might make it a little confusing to pull out to a helper function to prepare `$data` to send since here, we'd have to send many different variables to the function (`$northstar_id`, `$user`, `$values`, and `$run`) while in `dosomethign_rogue.cron.inc`, it gets these values from the table/from the `$task` object. I think it makes more sense to keep as is unless I'm not understanding correctly?

angaither replied:
@chloealee the reason I'd advocate for pulling it out is we are doing it in multiple places, and I will also be creating a similar object for the migration script. 
We can pass in an array or an object with all the variables and a common function can prepare it 

Similar to[ `dosomething_northstar_create_user`](https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_northstar/dosomething_northstar.module#L226)

chloealee replied:
@angaither ah okay makes sense - thank you for the example!

angaither commented on lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:
```diff
@@ -352,6 +352,25 @@ function _campaign_resource_reportback($nid, $values) {
+    $rbid = dosomething_reportback_exists($rogue_reportback['data']['campaign_id'], $rogue_reportback['data']['campaign_run_id'], $rogue_reportback['data']['drupal_id']);
+
+    // Store reference to the rb in rogue, redirect user to permalink page.
+    if ($rbid) {
```
what happens if there's no rbid?

chloealee replied:
would it make sense to do a watchdog error here? This would mean that Rogue's `POST` wouldn't have been successful.

angaither replied:
doesn't `rbid` not exist if it's not already saved to the phoenix DB?

I meant more what's the case where the transaction ID doesn't exist & the reportback id doesn't exist, doesn't that need to save the reportback?

chloealee replied:
oh yes I see what you mean now, @angaither ! So if there is no transaction ID, then it sends it to Rogue and Rogue sends back to this function and should now have a transaction id. Since it now has a transaction id, it'll move on to the `else` statement and then save.

chloealee committed:
adds comments for context for why we only send to rogue if there is no transaction id

chloealee committed:
adds helper function
:

angaither committed:
only create the field if the field doesn't already exist

noticed this error in the jenkins log
`Cannot add field dosomething_rogue_failed_task_log.northstar_id: field already exists.`

chloealee committed:
adds dosomething_rogue_transform_reportback helper and changes name of dosomething_rogue_format_crop_values helper

Commit:
allow reportback post to be an update with no image

angaither commented on lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:
```diff
@@ -338,8 +337,9 @@ function _campaign_resource_reportback($nid, $values) {
+  $values['file'] = dosomething_helpers_get_data_uri_from_fid($file->fid);
 
-  // @todo: Move this logic into dosomething_reportback_save().
+  // // @todo: Move this logic into dosomething_reportback_save().
```
double commented here

angaither commented on lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc:
```diff
@@ -352,6 +352,27 @@ function _campaign_resource_reportback($nid, $values) {
-  return dosomething_reportback_save($values, $user);
+  $transaction_id = $_SERVER['HTTP_X_REQUEST_ID'];
+
+  // If there is no transaction id, the request is coming from an app other than Rogue.
```
I'd add (mobile app, or sms reportbacks) here to be super clear in the comment

angaither commented on lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module:
```diff
@@ -263,3 +246,59 @@ function dosomething_rogue_rb_exists_in_rogue($rbid)
+ * @return array
+ */
+function dosomething_rogue_transform_reportback($northstar_id, $user, $values, $run)
+{
```
this bracket belongs on the row above in drupal, there's a few other times in this PR it's wrong

angaither commented on lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module:
```diff
@@ -263,3 +246,59 @@ function dosomething_rogue_rb_exists_in_rogue($rbid)
+/**
+ * Transform reportback into the appropriate scheme to send to Rogue.
+ *
+ * @param $northstar_id - User's Northstar user id.
```
you should probably only send one user value in, either ns id, or drupal user. I know we are starting to prefer NS id, so it's probably safe to only send that.

chloealee replied:
since there is a small percentage of people that do not have NS Ids, should I do a check before this helper function to send `$northstar_id` if there is one and if not, send the `$user`?

angaither commented on lib/modules/dosomething/dosomething_rogue/dosomething_rogue.module:
```diff
@@ -263,3 +246,59 @@ function dosomething_rogue_rb_exists_in_rogue($rbid)
+ * @param $northstar_id - User's Northstar user id.
+ * @param $user - Drupal user object.
+ * @param $values - Reportback values to send to Rogue.
+ * @param $run - Campaign run id.
```
the run should be a part of the values array like the other vars

angaither commented on lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc:
```diff
@@ -29,33 +29,37 @@
-        $values = [
+
+        $data = [
           'nid' => $task->campaign_id,
```
should this section be updated to use the transform function as well?

chloealee replied:
I was getting a little caught up on this because in order to send it to `dosomething_rogue_transform_reportback`, we'd be sending in only the `$task` object to fill out `$data` array and seemed to not match up exactly so thought it was cleaner just to keep as is?

angaither replied:
hmmm, I think it makes it a little less clear if there's multiple ways this data can be formatted though

chloealee replied:
ohhh okay I get what you are saying now - will change!

chloealee replied:
@angaither sorry - confusion again! Looking at this, I think it's okay to leave as is since it eventually will get to the transformer anyway? The flow for the cron job is that it gets sent to `dosomething_rogue_send_reportback_to_rogue` [here](https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc#L49), and in this function, it calls `dosomething_rogue_transform_reportback` so that data will just get formatted eventually anyway. LMK what you think!

chloealee replied:
@angaither i did do some tweaking that should make it better though so I'll push that up as well!

chloealee committed:
clarifies comment that it is coming from mobile app or SMS reportback

Commit:
update docs

chloealee committed:
refactor

angaither commented on lib/modules/dosomething/dosomething_rogue/dosomething_rogue.cron.inc:
```diff
@@ -29,33 +29,38 @@
-          'crop_height' => $task->crop_height,
-          'crop_rotate' => $task->crop_rotate,
         ];
+        // $data = dosomething_rogue_transform_reportback($task);
```
this line is still commented out?

chloealee replied:
oh right - I will delete that! Just had as a placeholder if we wanted to send to the helper function.

angaither replied:
maybe we are just going in circles here, but shouldn't that whole chunk be using the helper function?

chloealee replied:
@angaither omg YES sorry this took me so long to wrap my head around! All fixed :)

chloealee committed:
everything working

angaither commented on lib/modules/dosomething/dosomething_reportback/dosomething_reportback.forms.inc:
```diff
@@ -410,13 +410,17 @@ function dosomething_reportback_form_submit($form, &$form_state) {
-      dosomething_rogue_store_rogue_references($rbid, $fid, $rogue_reportback);
+      // If a file was added, store references to the relevant Rogue data
+      if (isset($file)) {
+        // $fid = array_pop($reportback->fids);
```
is this supposed to be commented out still?

chloealee committed:
fixes dosomething_rogue_format_crop_values helper

chloealee committed:
uses new helper method

chloealee committed:
?

chloealee committed:
gets rid of extra spaces

chloealee commented:
closing with fixes from #7227
